### ISSUE TITLE
OSM space too slow to create agents with large amounts of data

### ISSUE THREAD

**Issue by scepeda78** [JuliaDynamics/Agents.jl#490]
By the way, my simulation has to do with PV generation, and I'm interested in knowing the neighbours within a radius.

```julia
using Distributed
using Agents  
using DataFrames
using Parquet
using ExcelFiles
mapa=OpenStreetMapSpace("chile-latest.osm",use_cache=true) # (this takes 18.1 sec to load with cache, first time is like 5 min, 437618 roadways, 721204 intersections)

mutable struct generador <: AbstractAgent
    id::Int
    pos::Tuple{Int,Int,Float64}
end
model = ABM(generador, mapa)

pos=osm_intersection((-33.433619,-70.664926),model)
human = generador(1,pos)
add_agent_pos!(human, model) (0.7 secs )

pos=osm_road((-33.433619,-70.664926),model) (0.9 secs)
human = generador(2,pos)
add_agent_pos!(human, model)

# this was my attemp to make it distribuited
#Cargar terrenos
archivo="terrenos_sin_BBCC_filtrado.parquet"
terrenos=DataFrame(read_parquet(archivo))
first(terrenos,5)
terrenos_seleccionados=terrenos[terrenos.region_ISO .== "AP",:]
#Crear agentes primer intento
using ProgressMeter
N = size(terrenos_seleccionados,1)
p = Progress(N);
update!(p,0)
jj = Threads.Atomic{Int}(0)
l = Threads.SpinLock()

Threads.@threads for id in 1:size(terrenos_seleccionados,1)
    pos=osm_random_road_position(model)
    human = generador(id,pos)
    add_agent_pos!(human, model)
    Threads.atomic_add!(jj, 1)
    Threads.lock(l)
    update!(p, jj[])
    Threads.unlock(l) 
end
```
(here the ETA time was 2 hours for 4 million households)
I m using a 32 core ubuntu 20 computer, and using 25 cores for the distributed process...

`

_Originally posted by @scepeda78 in https://github.com/JuliaDynamics/Agents.jl/issues/403#issuecomment-819637673_

**Comment by Libbum** [comment]
This will be difficult to test without `chile-latest.osm` and `terrenos_sin_BBCC_filtrado.parquet`. Are these files available as links somewhere?

**Comment by scepeda78** [comment]
> This will be difficult to test without `chile-latest.osm` and `terrenos_sin_BBCC_filtrado.parquet`. Are these files available as links somewhere?

https://1drv.ms/u/s!AkIvnTxgic15hOFD_Og9ooxC7pyBsQ?e=yOjmBo

**Comment by Libbum** [comment]
So this is not too surprising now that I've run some numbers, but there are perhaps some discrepancies.

On my machine:
```julia
julia> size(terrenos_seleccionados,1) 
45833
julia> @btime OSM.random_road_position($model)
  32.333 ms (35 allocations: 1.28 KiB)
```
That looks like 46,000 not 4 million agents, correct? 

`(0.032633*45833)/60 = 24.9 minutes` to get the road positions, which is not great but also not crazy&mdash;certainly not 2 hours.. 

Without using threads
```julia
@time for id in 1:size(terrenos_seleccionados,1)
    pos=osm_random_road_position(model)
    human = generador(id,pos)
    add_agent_pos!(human, model)
end
1538.163697 seconds (1.88 M allocations: 71.860 MiB, 0.00% gc time, 0.00% compilation time)
# 25.6 minutes
```

My assertion here is that your thread spool up time & lock times are really the issue in terms of this 2 hour number. 

We do want to make the OSM space more performant, it's on the roadmap for Agents.jl, but there is a lot of underlying work to get this fixed up, so for the moment my suggestion is to not use threads for the loading. We'll work on decreasing these load times as best we can in the future.

**Comment by scepeda78** [comment]
Hi, my mistake, I had to slice the full load to a smaller region to do some tests. The full load is under the terrenos Dataframe.
The ProgressMeter and the Threads Atomic was to have an idea of the full run, without have to wait the two full hours to add the agents. It seems the search of roads or intersections is the expensive process under the other library OpenStreetMapX,since if I have a smaller osm file, to add 45000 agents took very small time. So maybe, under that library it can be speed up the search ... (for example nearest_node and nodes_within_range, has a for over the full nodes of the graph it seems).

```julia
using ProgressMeter
N = size(terrenos,1)
p = Progress(N);
update!(p,0)
jj = Threads.Atomic{Int}(0)
l = Threads.SpinLock()

Threads.@threads for id in 1:size(terrenos,1)
    pos=osm_random_road_position(model)
    human = generador(id,pos)
    add_agent_pos!(human, model)
    Threads.atomic_add!(jj, 1)
    Threads.lock(l)
    update!(p, jj[])
    Threads.unlock(l) 
end
```

**Comment by Libbum** [comment]
OK, fair enough. In that case your threading is definitely helping.

However,  `osm_random_road_position` is NOT possible using OpenStreetMapX, neither are searches over roads. Those are additional features that Agents.jl provides.

You could swap out `osm_random_road_position(model)` and use `random_position(model)`. This will be quick and the same as what OpenStreetMapX can provide you. However: this is not really what you're after for your map I believe. 

`nearest_node` is equivalent to "Nearest intersection". The underlying graph is only intersection based and therefore if you add agents in that manner you'll see many agents with nearest distance of zero, since they will sit on the same intersection, not some distance down some road. These distinctions are [in the documentation](https://juliadynamics.github.io/Agents.jl/stable/api/#OpenStreetMap-space-exclusives), but perhaps are not clear enough.

In other words, OpenStreetMapX will provide you a discrete map, looking only at distances between intersections, with all of your agents sitting at intersections and no-where else. Agents.jl can provide you with true, continuous locations and distances, but at a computational cost.

If you can deal with the inaccuracy of OpenStreetMapX, but still wish to use Agents, just use `random_position(model)`.

**Comment by scepeda78** [comment]
> OK, fair enough. In that case your threading is definitely helping.
> 
> However, `osm_random_road_position` is NOT possible using OpenStreetMapX, neither are searches over roads. Those are additional features that Agents.jl provides.
> 
> You could swap out `osm_random_road_position(model)` and use `random_position(model)`. This will be quick and the same as what OpenStreetMapX can provide you. However: this is not really what you're after for your map I believe.
> 
> `nearest_node` is equivalent to "Nearest intersection". The underlying graph is only intersection based and therefore if you add agents in that manner you'll see many agents with nearest distance of zero, since they will sit on the same intersection, not some distance down some road. These distinctions are [in the documentation](https://juliadynamics.github.io/Agents.jl/stable/api/#OpenStreetMap-space-exclusives), but perhaps are not clear enough.
> 
> In other words, OpenStreetMapX will provide you a discrete map, looking only at distances between intersections, with all of your agents sitting at intersections and no-where else. Agents.jl can provide you with true, continuous locations and distances, but at a computational cost.
> 
> If you can deal with the inaccuracy of OpenStreetMapX, but still wish to use Agents, just use `random_position(model)`.

In my model what I care is to add agents in the continuos space, but the examples in the web page are related to add them into the roads. Also, I care the distance between them, since I will model the influence of others in the purchase of a PV system. Therefore, is there an example adding agents in the continuos space and get nearby agents based on the distance?.

**Comment by Datseris** [comment]
just use `random_position(model)` when you add the agents. `nearby_agents` always uses real distance in meters and takes into account road length, doesn't matter if the yare in an intersection or not.

**Comment by Libbum** [comment]
> doesn't matter if the yare in an intersection or not.

True but that doesn't solve the actual issue. Agents will be always nearby since they cannot reside in the middle of some road in this case.

@scepeda78: Are your agents ever going to move once you have loaded this data? If not, I would say that Agents.jl is not the most performant solution for your model.

Nearest neighbors is a complicated algorithm and is very slow in the case of moving agents. We must check the distance of all agent pairs every time something changes. For 4 million pairs that is 16000000000000 checks each step! If your agents are going to stay where they are over your simulation, the better solution would be to look at using a KDTree from [NearestNeighbors.jl](https://github.com/KristofferC/NearestNeighbors.jl) to do this search.

**Comment by scepeda78** [comment]
> > doesn't matter if the yare in an intersection or not.
> 
> True but that doesn't solve the actual issue. Agents will be always nearby since they cannot reside in the middle of some road in this case.
> 
> @scepeda78: Are your agents ever going to move once you have loaded this data? If not, I would say that Agents.jl is not the most performant solution for your model.
Yes, they will stay fixed the whole simulation, since they are households.
> 
> Nearest neighbors is a complicated algorithm and is very slow in the case of moving agents. We must check the distance of all agent pairs every time something changes. For 4 million pairs that is 16000000000000 checks each step! If your agents are going to stay where they are over your simulation, the better solution would be to look at using a KDTree from [NearestNeighbors.jl](https://github.com/KristofferC/NearestNeighbors.jl) to do this search.
Thanks I'll look into that!

**Comment by Datseris** [comment]
Given that we now use internally LightOSM, massively improving performance over OpenStreetMapX, I'm closing this. Feel free to open **a new issue** if a similar problem remains.
